mir-ci/mir_ci/robot_libraries/Screencopy.py

```diff
@@ -0,0 +1,169 @@
+import asyncio
+import base64
+import os
+import time
+from io import BytesIO
+from typing import List
+
+from mir_ci.screencopy_tracker import ScreencopyTracker
+from PIL import Image
+from robot.api import logger
+from robot.api.deco import keyword, library
+from RPA.Images import Images
+from RPA.recognition.templates import ImageNotFoundError
+
+
+@library(scope="GLOBAL")
+class Screencopy(ScreencopyTracker):
+    """
+    A Robot Framework library for capturing screenshots from the
+    Wayland display and performing image template matching.
+
+    The client connects to the display upon entering the first keyword,
+    and disconnects when the library goes out of scope.
+
+    If WAYLAND_DISPLAY is not defined, it defaults to 'wayland-0'.
+    """
+
+    ROBOT_LISTENER_API_VERSION = 3
+    TOLERANCE = 0.8
+
+    def __init__(self) -> None:
+        self.ROBOT_LIBRARY_LISTENER = self
+        display_name = os.environ.get("WAYLAND_DISPLAY", "wayland-0")
+        super().__init__(display_name)
+        self._rpa_images = Images()
+
+    @keyword
+    async def test_match(self, template: str):
+        """
+        Grab a screenshot and check for a match with the given template.
+
+        :param template: path to an image file to be used as template
+        :return: list of matched regions
+        :raises ImageNotFoundError: if no match is found
+        """
+        await self.connect()
+        screenshot = None
+
+        try:
+            screenshot = self.grab_screenshot()
+            regions = self._rpa_images.find_template_in_image(
+                screenshot,
+                template,
+                tolerance=self.TOLERANCE,
+            )
+
+            return [
+                {
+                    "left": region.left,
+                    "top": region.top,
+                    "right": region.right,
+                    "bottom": region.bottom,
+                }
+                for region in regions
+            ]
+        except (RuntimeError, ValueError, ImageNotFoundError) as exc:
+            self._log_failed_match(screenshot, template)
+            raise ImageNotFoundError from exc
+
+    @keyword
+    async def wait_match(self, template: str, timeout: int = 10) -> List[dict]:
+        """
+        Grab screenshots and compare until there's a match with the provided
+        template.
+
+        :param template: path to an image file to be used as template
+        :param timeout: timeout in seconds
+        :return: list of matched regions
+        :raises ImageNotFoundError: if no match is found within the timeout
+        """
+        await self.connect()
+        regions = []
+        last_frame_count = 0
+        matched = False
+        screenshot = None
+        end_time = time.time() + float(timeout)
+        while time.time() < end_time:
+            if self.frame_count == last_frame_count:
+                await asyncio.sleep(0)
+            else:
+                last_frame_count = self.frame_count
```

Comment:
Shouldn't we have that dance in `grab_screenshot`? I mean, if we don't have a frame, we shouldn't be grabbing it. And in that case, don't we need a timeout on `grab_screenshot`? And couldn't we `await` it?

Feel free to log this as an issue to tackle later, won't block on it.

Reply:
Okay. I need to think more about it. `grab_screenshot` returns the current frame available. Maybe it would change its  behavior to wait for the next frame, which is not what `Test Match` expects.

Reply:
It shouldn't matter, though? I mean, if we skip over the current frame, the next one should be just as appropriate for `Test Match`, otherwise the test would be unpredictable?

But also, there's no reason for `grab_screenshot` to grab the current frame, _unless_ it's the same frame it's already seen - in which case it should wait for the next one (and we'd need a `await` with a timeout here).

Reply:
So the only case where this may get problematic is when the app/compositor stops drawing. So then two consecutive `grab_screenshot`s can either: a) grab the same screenshot b) timeout on the second call.

I _think_ b) is a very special case (requiring a new grab) that we don't have to tackle right now, and we can just go with a).

I think we can drop this dance altogether, `await asyncio.wait_for(self.grab_screenshot(), timeout)` and we'll timeout both when no screenshot ever picked up and when no match was found.